overlord/snapstate/snapmgr.go

```diff
@@ -404,28 +399,45 @@ func (m *SnapManager) ensureRefreshes() error {
 
 	tr := config.NewTransaction(m.state)
 
-	// allow disabling auto-refresh in the tests
-	if osutil.GetenvBool("SNAPD_DEBUG") {
-		var refreshDisabled bool
-		err := tr.Get("core", "refresh.disabled", &refreshDisabled)
-		if err != nil && !config.IsNoOption(err) {
-			return err
-		}
-		if refreshDisabled {
-			return nil
-		}
+	// get last refresh time
+	var lastRefresh time.Time
+	err := m.state.Get("last-refresh", &lastRefresh)
+	if err != nil && err != state.ErrNoState {
+		return err
 	}
 
-	var lastRefresh time.Time
-	err := tr.Get("core", "refresh.last", &lastRefresh)
+	// get our schedule
+	refreshScheduleStr := defaultRefreshSchedule
+	err = tr.Get("core", "refresh.schedule", &refreshScheduleStr)
 	if err != nil && !config.IsNoOption(err) {
 		return err
 	}
+	refreshSchedule, err := timeutil.ParseSchedule(refreshScheduleStr)
+	if err == nil {
+		err = refreshScheduleUsesWeekdays(refreshSchedule)
+	}
+	if err != nil {
+		logger.Noticef("cannot use refresh.schedule: %s", err)
+		refreshSchedule, err = timeutil.ParseSchedule(defaultRefreshSchedule)
+		if err != nil {
+			panic(fmt.Sprintf("defaultRefreshSchedule cannot be parsed: %s", err))
+		}
+		tr.Set("core", "refresh.schedule", defaultRefreshSchedule)
+		tr.Commit()
+	}
 
-	nextRefresh := lastRefresh.Add(minRefreshInterval).Add(m.refreshRandomness)
-	if time.Now().Before(nextRefresh) {
-		return nil
+	// already have a refresh timer
+	if m.nextRefresh != nil {
+		if m.currentRefreshSchedule == refreshScheduleStr {
+			return nil
+		}
+		// the refresh schedule has changed
+		logger.Debugf("Option refresh-schedule changed, reloading.")
+		m.nextRefresh.Stop()
+		m.nextRefresh = nil
+		m.lastRefreshAttempt = time.Time{}
```

Comment:
I'm also not sure 100% why we need the timer? wouldn't checking if the nextRefresh time is in the past be enough and try to do the change creation job here directly?